src/rust_option.rs

```diff
@@ -0,0 +1,230 @@
+#![allow(missing_docs)]
+
+use core::mem::ManuallyDrop;
+use std::boxed::Box;
+use std::ops::Deref;
+use std::pin::Pin;
+
+#[repr(C)]
+union OptionInner<T> {
+    value: ManuallyDrop<T>,
+    empty: usize,
+}
+
+// ABI compatible with C++ rust::Option<T> (not necessarily core::option::Option<T>).
+#[repr(C)]
+pub struct RustOption<T> {
+    inner: OptionInner<T>,
+}
+
+impl<T> RustOption<T> {
+    pub fn new() -> Self {
+        RustOption {
+            inner: OptionInner { empty: 0 },
+        }
+    }
+
+    pub fn value(&self) -> Option<&T> {
+        if unsafe { self.inner.empty != 0 } {
+            unsafe { Some(self.inner.value.deref()) }
+        } else {
+            None
+        }
+    }
+
+    pub unsafe fn into_inner_unchecked(mut self) -> T {
+        ManuallyDrop::take(&mut self.inner.value)
+    }
+}
+
+impl<'a, T> RustOption<&'a T> {
```

Comment:
I think some of these implementations can be removed but maybe its future proofing to have them? I can try to see which ones are unneeded